### PR TITLE
Drop wait from ContextualServer and other cleanups

### LINKED ISSUE #442: Make port number dynamic
Currently SonarTS server is started on port 55555

### LINKED ISSUE #445: Stop using InputFile.absolutePath()
This method is deprecated

### LINKED ISSUE #446: Drop wait from ContextualServer.start()

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #446 
Fixes #442 
Fixes #445

### CODE REVIEW

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java [hunk 1]
```diff
@@ -79,57 +76,49 @@ private void startSonarTSServer() {
...
-
-    } catch (InterruptedException e) {
-      LOG.error("Failed to wait until the SonarTS Server is up", e);
-      throw new RuntimeException(e);
     }
   }
 
+  private static void tryToConnect() throws IOException {
+    InetSocketAddress address = new InetSocketAddress("localhost", 55555);
+    Socket socket = new Socket();
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java]: Bug Bug: Use try-with-resources or close this "Socket" in a "finally" clause. ([squid:S2095](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2095))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-446&issues=AWEDyrk3f6JiG_HArpQ8&open=AWEDyrk3f6JiG_HArpQ8)

[ISSUE_KEY]: AWEDyrk3f6JiG_HArpQ8

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java [hunk 2]
```diff
@@ -58,78 +69,75 @@ public ContextualServer(Configuration configuration, ExecutableBundleFactory bun
...
-      // e.g. loop with accessing the port
-      synchronized (this) {
-        wait(1000);
-      }
+      serverSocket = new ServerSocket(0);
+      serverSocket.setSoTimeout(5_000);
+      ProcessBuilder processBuilder = new ProcessBuilder(bundle.getSonarTSServerCommand(serverSocket.getLocalPort()).commandLineTokens());
+      setNodePath(processBuilder);
+      serverProcess = processBuilder.start();
+      socket = serverSocket.accept();
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java]: Code Smell Code Smell: Synchronize this lazy initialization of 'socket' ([squid:S2444](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2444))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-446&issues=AWEFRU3REFfjxkQcQHBy&open=AWEFRU3REFfjxkQcQHBy)

[ISSUE_KEY]: AWEFRU3REFfjxkQcQHBy

Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java]: Code Smell Code Smell: Make the enclosing method "static" or remove this set. ([squid:S2696](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2696))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-446&issues=AWEFRU3REFfjxkQcQHBx&open=AWEFRU3REFfjxkQcQHBx)

[ISSUE_KEY]: AWEFRU3REFfjxkQcQHBx

File: sonarts-core/src/runner/sonartsServer.ts [hunk 3]
```diff
@@ -63,19 +63,14 @@ export function start(port: number): Promise<{ server: net.Server; port: number
...
-    });
-  });
-  return new Promise(resolve => {
-    server.listen(port, "localhost", () => {
-      console.log(`SonarTS Server started on port ${port} from folder ${__dirname}`);
-      resolve({ server, port });
+      client.write(JSON.stringify(issues));
     });
   });
+  return new Promise(resolve => resolve(client));
```
Comment (vilchik-elena) [inline sonarts-core/src/runner/sonartsServer.ts]: @saberduck fix this promise

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java [hunk 4]
```diff
@@ -58,78 +69,75 @@ public ContextualServer(Configuration configuration, ExecutableBundleFactory bun
...
     } else {
-      LOG.warn("Failed to stop SonarTS Server");
+      LOG.warn("SonarTS Server was already stopped");
     }
   }
+
+  private void terminate() {
+    try {
+      IOUtils.closeQuietly(serverSocket);
+      IOUtils.closeQuietly(socket);
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java]: @saberduck change order

File: sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java [hunk 5]
```diff
@@ -19,13 +19,137 @@
...
+  public JUnitTempFolder temp = new JUnitTempFolder();
+
+  @Rule
+  public final LogTester logTester = new LogTester();
+
+  /**
+   * First try 'node' from frontend-maven-plugin, fallback to 'node' from the path
+   */
+  @BeforeClass
+  public static void setUp() {
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java]: @saberduck can we drop the duplication with main sensor test?

File: sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java [hunk 6]
```diff
@@ -19,13 +19,137 @@
...
+    contextualServer.start();
+    assertThat(logTester.logs(LoggerLevel.ERROR)).containsOnlyOnce("Failed to start SonarTS Server");
+    assertThat(contextualServer.isAlive()).isFalse();
+  }
+
+  private TestBundleFactory mockTSServer() {
+    return new TestBundleFactory().command(resourceScript("/mockSonarTSServer.js"));
+  }
+
+  private static class TestBundleFactory implements ExecutableBundleFactory {
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java]: @saberduck can we drop the duplication with main sensor test?

File: sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java [hunk 7]
```diff
@@ -19,13 +19,137 @@
...
+import org.sonar.api.utils.log.LogTester;
+import org.sonar.api.utils.log.LoggerLevel;
+import org.sonar.plugin.typescript.executable.ExecutableBundle;
+import org.sonar.plugin.typescript.executable.ExecutableBundleFactory;
+import org.sonar.plugin.typescript.executable.SonarTSCommand;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.sonar.plugin.typescript.TestUtils.resourceScript;
 
 public class ContextualServerTest {
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java]: @saberduck let's add unit test for sensor as well

Comment (saberduck) [reply sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java]: let's postpone this

File: sonarts-core/src/runner/sonartsServer.ts [hunk 8]
```diff
@@ -21,63 +21,70 @@ import * as net from "net";
...
-
-      const program = service.getProgram();
-      const sourceFile = program.getSourceFile(file);
-      if (!sourceFile) {
-        console.error(`No SourceFile found for file ${file} with configuration ${tsConfig}`);
-        socket.write(EMPTY_ANSWER);
-        return;
+export function start(port: number) {
+  const client = net.createConnection(port, "localhost", () => {
+    console.log("sonartsServer connected to " + port);
```
Comment (vilchik-elena) [inline sonarts-core/src/runner/sonartsServer.ts]: @saberduck `SonarTS Server`

**Pull request merged**